File: C++/affine/AffineCryptanalysis.cpp
#include <tuple>
#include "Affine.h"
#include "../caesarcryptanalysis/CaesarCryptanalysis.h"

using namespace std;

tuple<int, int> getKey(string& ciphertext) {
    int a = 0;
    int b = 0;
    tuple<float, tuple<int,int>> best = make_tuple(1000, tuple<int, int>(0,0));
    for (const auto& pair : inverses) {
        int i = pair.first; // Access the key from the map
        for (int j = 0; j < 26; j++) {
            string curr = decryptAffine(ciphertext, i, j);
            float dist = 0.065 - getMIC(curr, true);
            if (dist < 0) {
                dist = -dist;
            }
            if (dist < get<0>(best)) {
                best = make_tuple(dist, tuple<int, int>(i,j));
            }
        }
    }
    return get<1>(best);
}

string decryptAffine(string& ciphertext) {
    tuple<int, int> key = getKey(ciphertext);
    return decryptAffine(ciphertext, get<0>(key), get<1>(key));
}

int main() {
    initializeInverses();
    string ciphertext = "PFIMKZZIXWQIDFWJIMDJWPPEQAARWVWXIZITAXAPSXTWJMPKXTPFWUKPFNWFIXTIPKPPIUWMNSPIZWWRNKTNWQKSMWRKMPPIUWKRAPAZDWADRWTITKRAPAZPWBPKXTIZWWRRIOWITITXPTAKRAPAZPWBPMAXAYIUTAIXCKRAPPAUKOWMSJWIPMWXASCFNWQKSMWEASMKITIPMMSDDAMWTPANWKQASDRWFSXTJWTQFKJKQPWMYFIQFUKOWMMWXMWNWQKSMWEASDJANKNREXWWTKRAPAZQFKJKQPWJMIXAJTWJPANWKNRWPATWQJEDPIPWMDWIQKRREYIPFKXWYQIDFWJIPMDJANKNRENWPPWJPAFKVWYKEUAJWQFKJKQPWJMKXTUKOWIPWKMIWJAXYFAWVWJIMKPPWUDPIXCPATWQJEDPIPNWQKSMWIPMKYFARWXWYQIDFWJKXTIPQKXNWJWKRREQAXZSMIXCKPPFWMPKJPPAKPPWUDPPAZICSJWASPYFKPPAWVWXTAKXTYFWJWPAWVWXNWCIX";
    string decrypted = decryptAffine(ciphertext);
    cout << "Decrypted text: " << decrypted << endl;
    cout << "Key: " << get<0>(getKey(ciphertext)) << ", " << get<1>(getKey(ciphertext)) << endl;
    return 0;
}

File: javaCryptography/src/affine/AffineCryptanalysis.java
package affine;
import caesar.*;
import tools.Tuple;

public class AffineCryptanalysis {

    public static Tuple<Integer, Integer> getKey(String ciphertext) {
        Tuple<Double, Tuple<Integer, Integer>> best = new Tuple<>((Double) 1000.0, new Tuple<>((Integer) 0, (Integer) 0));
        for (int i : Affine.inverses.keySet()) {
            for (int j = 0; j < 26; j++) {
                String curr = Affine.decryptAffine(ciphertext, i, j);
                double mic = CaesarCryptanalysis.getMIC(curr, true);
                double dist = Math.abs(0.065 - mic);
                if (dist < best.getFirst()) {
                    best.setFirst(dist);
                    best.setSecond(new Tuple<>(i, j));
                }
            }
        }
        return best.getSecond();
    }
    

    public static String decryptAffineNoKey(String ciphertext) {
    	Tuple<Integer, Integer> key = getKey(ciphertext);
    	return Affine.decryptAffine(ciphertext, key.getFirst(), key.getSecond());
    }
    
    public static void main(String[] args) {
    	String cipher = "PFIMKZZIXWQIDFWJIMDJWPPEQAARWVWXIZITAXAPSXTWJMPKXTPFWUKPFNWFIXTIPKPPIUWMNSPIZWWRNKTNWQKSMWRKMPPIUWKRAPAZDWADRWTITKRAPAZPWBPKXTIZWWRRIOWITITXPTAKRAPAZPWBPMAXAYIUTAIXCKRAPPAUKOWMSJWIPMWXASCFNWQKSMWEASMKITIPMMSDDAMWTPANWKQASDRWFSXTJWTQFKJKQPWMYFIQFUKOWMMWXMWNWQKSMWEASDJANKNREXWWTKRAPAZQFKJKQPWJMIXAJTWJPANWKNRWPATWQJEDPIPWMDWIQKRREYIPFKXWYQIDFWJIPMDJANKNRENWPPWJPAFKVWYKEUAJWQFKJKQPWJMKXTUKOWIPWKMIWJAXYFAWVWJIMKPPWUDPIXCPATWQJEDPIPNWQKSMWIPMKYFARWXWYQIDFWJKXTIPQKXNWJWKRREQAXZSMIXCKPPFWMPKJPPAKPPWUDPPAZICSJWASPYFKPPAWVWXTAKXTYFWJWPAWVWXNWCIX";
    	System.out.println(decryptAffineNoKey(cipher));
    }
}
